shared/js/background/declarative-net-request.js

```diff
@@ -262,36 +279,42 @@ async function updateExtensionConfigRules (etag = null, configValue = null) {
  * @returns {Promise}
  */
 export async function onConfigUpdate (configName, etag, configValue) {
-    // TDS (aka the block list).
-    if (configName === 'tds') {
-        const extensionVersion = browserWrapper.getExtensionVersion()
-        const [ruleIdStart] = ruleIdRangeByConfigName[configName]
-        const latestState = { etag, extensionVersion }
-        if (!(await configRulesNeedUpdate(configName, latestState))) {
-            return
-        }
-
-        await startup.ready()
-        // @ts-ignore: Once startup.ready() has finished, surrogateList will be
-        //             assigned.
-        const supportedSurrogates = new Set(Object.keys(trackers.surrogateList))
-
-        const {
-            ruleset, matchDetailsByRuleId
-        } = await generateTdsRuleset(
-            configValue,
-            supportedSurrogates,
-            '/web_accessible_resources/',
-            chrome.declarativeNetRequest.isRegexSupported,
-            ruleIdStart + 1
-        )
-
-        await updateConfigRules(configName, latestState, ruleset, matchDetailsByRuleId)
+    const extensionVersion = browserWrapper.getExtensionVersion()
+    // Run an async lock on all blocklist updates so the latest update is always processed last
+    ruleUpdateLock = ruleUpdateLock.then(async () => {
```

Comment:
This implements a serial queue for updates, so all changes to tds and config must be completed before the next update is processed. This prevents out of order calls to `updateDynamicRules` leaving the rules in an arbitrary state.

Reply:
The issue I had is that if the tds list update happens twice in quick succession, it is not guaranteed that your dynamic rules are those generated from the last list update. This could be because when the TDS rule generation hits a regex rule, it has to do an async call. That tick then allows the next update that sitting on the event queue to run and generate rules before the first update finishes.

Reply:
Ah right, got it.